# Shwet Jain



## Assignment 4
### Sentiment Analysis


### Importing all the necessary libraries

In [168]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
stop_words = set(stopwords.words('english'))
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_recall_fscore_support
from sklearn.svm import LinearSVC
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LogisticRegression
from nltk.stem import PorterStemmer 

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\shwet\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


### Reading the Data into the dataframe "df_sentiment"

In [169]:
df_sentiment = pd.read_csv('/Users/shwet/OneDrive/Desktop/DA_4/HW4_yelp_sentiment.csv')
print(df_sentiment.shape)

(988, 2)


In [171]:
df_sentiment.head()

reviews sentiment
0                           Wow... Loved this place.         p
1                                 Crust is not good.         n
2          Not tasty and the texture was just nasty.         n
3  Stopped by during the late May bank holiday of...         p
4  The selection on the menu was great and so wer...         p

### TfidfVectorizer combines all the options of CountVectorizer and TfidfTransformer in a single model
### Using tf-idf vectorizer to change text to feature
### Getting rid of all the stop words
### An ngram_range of (1, 1) means only unigrams, (1, 2) means unigrams and bigrams, and (2, 2) means only bigrams 
### We are using unigrams and bigrams i.e (1, 2)
### Converting into lower-case

In [172]:
sent_rev = df_sentiment.reviews

vect = TfidfVectorizer(stop_words = stop_words, ngram_range =(1,2), min_df = 2, lowercase=True)

vect.fit(sent_rev)

TfidfVectorizer(analyzer=u'word', binary=False, decode_error=u'strict',
        dtype=<type 'numpy.float64'>, encoding=u'utf-8', input=u'content',
        lowercase=True, max_df=1.0, max_features=None, min_df=2,
        ngram_range=(1, 2), norm=u'l2', preprocessor=None, smooth_idf=True,
        stop_words=set([u'all', u'just', u"don't", u'being', u'over', u'both', u'through', u'yourselves', u'its', u'before', u'o', u'don', u'hadn', u'herself', u'll', u'had', u'should', u'to', u'only', u'won', u'under', u'ours', u'has', u"should've", u"haven't", u'do', u'them', u'his', u'very', u"you've", u..., u'a', u'off', u'i', u'm', u'yours', u"you'll", u'so', u'y', u"she's", u'the', u'having', u'once']),
        strip_accents=None, sublinear_tf=False,
        token_pattern=u'(?u)\\b\\w\\w+\\b', tokenizer=None, use_idf=True,
        vocabulary=None)

In [173]:
sent_rev_tfidf = vect.transform(sent_rev)

In [256]:
sent_rev_tfidf

<988x968 sparse matrix of type '<type 'numpy.float64'>'
	with 4812 stored elements in Compressed Sparse Row format>

## We are now forming a Document Term Matrix

In [46]:
dtm = pd.DataFrame(sent_rev_tfidf.toarray(), columns=vect.get_feature_names())


In [47]:
dtm.head()

10  10 minutes  100   12   20  20 minutes   30  30 min  30 minutes   35  \
0  0.0         0.0  0.0  0.0  0.0         0.0  0.0     0.0         0.0  0.0   
1  0.0         0.0  0.0  0.0  0.0         0.0  0.0     0.0         0.0  0.0   
2  0.0         0.0  0.0  0.0  0.0         0.0  0.0     0.0         0.0  0.0   
3  0.0         0.0  0.0  0.0  0.0         0.0  0.0     0.0         0.0  0.0   
4  0.0         0.0  0.0  0.0  0.0         0.0  0.0     0.0         0.0  0.0   

   ...       wow  wrap  wrong  year  years  years ago  yet  yummy  zero  \
0  ...  0.572012   0.0    0.0   0.0    0.0        0.0  0.0    0.0   0.0   
1  ...  0.000000   0.0    0.0   0.0    0.0        0.0  0.0    0.0   0.0   
2  ...  0.000000   0.0    0.0   0.0    0.0        0.0  0.0    0.0   0.0   
3  ...  0.000000   0.0    0.0   0.0    0.0        0.0  0.0    0.0   0.0   
4  ...  0.000000   0.0    0.0   0.0    0.0        0.0  0.0    0.0   0.0   

   zero stars  
0         0.0  
1         0.0  
2         0.0  
3         0.0  
4         0.0  

[5 rows x 968 columns]

In [51]:
X = dtm.values
y = df['sentiment'].values

### Splitting of Data into Training and Testing set 

In [103]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=100)
print(X_train.shape, y_train.shape)
print(X_test.shape,y_test.shape)

((790L, 968L), (790L,))
((198L, 968L), (198L,))


## Logistic Regression


## Building the model using (.fit) the train data and then finding the accuracy on the test data

## Trying different values of c to find the highest accuracy

In [132]:
from sklearn.metrics import accuracy_score
for c in [0.01,0.25,0.5,1,5,7,9,10,13,15,20,22,25]:
    
    LR = LogisticRegression(C=c)
    LR.fit(X_train, y_train)
    print ("Accuracy for C=%s: %s" 
           % (c, accuracy_score(y_test, LR.predict(X_test))))

Accuracy for C=0.01: 0.51010101010101
Accuracy for C=0.25: 0.7424242424242424
Accuracy for C=0.5: 0.7525252525252525
Accuracy for C=1: 0.7575757575757576
Accuracy for C=5: 0.7525252525252525
Accuracy for C=7: 0.7474747474747475
Accuracy for C=9: 0.7474747474747475
Accuracy for C=10: 0.7474747474747475
Accuracy for C=13: 0.7575757575757576
Accuracy for C=15: 0.7575757575757576
Accuracy for C=20: 0.7626262626262627
Accuracy for C=22: 0.7626262626262627
Accuracy for C=25: 0.7525252525252525


## Trying alternate values of the loss functions and trial error of other parameters 

In [133]:
classifier = LogisticRegression(class_weight= 'balanced', penalty= 'l2',C=25, max_iter=100)
classifier.fit(X_train, y_train)
score = classifier.score(X_test, y_test)
print('Accuracy is' + ' ' + str(score*100) + '%')

Accuracy is 76.76767676767676%


## Confusion Matrix

In [136]:
predict = classifier.predict(X_test)
print("Confusion matrix:")
print(confusion_matrix(y_test, predict))
print("Report:")
from sklearn.metrics import classification_report
print(classification_report(y_test, predict))

Confusion matrix:
[[70 19]
 [27 82]]
Report:
              precision    recall  f1-score   support

           n       0.72      0.79      0.75        89
           p       0.81      0.75      0.78       109

   micro avg       0.77      0.77      0.77       198
   macro avg       0.77      0.77      0.77       198
weighted avg       0.77      0.77      0.77       198



## Support Vector Machine

In [125]:
from sklearn import svm
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score, ShuffleSplit
from sklearn import metrics

In [126]:
lin_clf = svm.LinearSVC(max_iter=100000)

## Different values of C

In [127]:
param_grid = {'C':[0.1,0.25,0.5,0.75,1]}

## GridSearch using cross-validation

In [128]:
from time import time
start = time()
grid_lsvm = GridSearchCV(lin_clf,param_grid,cv=5,scoring="accuracy")
grid_lsvm.fit(X_train,y_train)
print("Linear SVM took %.2f seconds"
       % (time() - start))

Linear SVM took 0.29 seconds


## Best Parameter for parameter C 

In [137]:
grid_lsvm.best_params_

{'C': 0.1}

In [139]:
grid_pred = grid_lsvm.predict(X_test)
acc= round(metrics.accuracy_score(y_test, grid_pred)*100, 2)
print ("Accuracy: " + str(acc) + "%")

Accuracy: 74.75%


## Confusion Matrix

In [140]:
print("Confusion Matrix for Random Forest Classifier:")
print('\n')
print(confusion_matrix(y_test,grid_pred))
print("Classification Report:")
print('\n')
print(classification_report(y_test, grid_pred))

Confusion Matrix for Random Forest Classifier:


[[76 13]
 [37 72]]
Classification Report:


              precision    recall  f1-score   support

           n       0.67      0.85      0.75        89
           p       0.85      0.66      0.74       109

   micro avg       0.75      0.75      0.75       198
   macro avg       0.76      0.76      0.75       198
weighted avg       0.77      0.75      0.75       198



## Neural Networks - ANN

In [257]:
X_train, X_test, y_train, y_test = train_test_split(sent_rev_tfidf, df_sentiment.sentiment, test_size=0.25, random_state=1000)

## I tried different activation functions with different hidden layer sizes to get the best accuracy  

In [262]:
nn = MLPClassifier(solver='adam', alpha=1e-5, hidden_layer_sizes = (180, 170, 150, 100),
                   activation = 'relu',learning_rate='adaptive',random_state=1, max_iter = 1500)

In [266]:
abc1= nn.fit(X_train,y_train)
prediction = abc1.predict(X_test)
tn,fp,fn,tp = confusion_matrix(y_test,prediction).ravel()
print(tn)
print(fp)
print(fn)
print(tp)

102
19
35
91


#### As this is Python 2, there's an exception for division

In [254]:
from __future__ import division

In [267]:
print( "Accuracy = " + str((tn+tp)/(tn+tp+fp+fn)))

Accuracy = 0.7813765182186235


# Logistic Regression - Accuracy: 76.67% 

# ANN - Accuracy: 78.13%

# SVM - Accuracy: 74.75%

## Out of the three models, Artificial Neural Netwrok is the model which gives highest accuracy. Also, ANN takes a lot more time (more computations as it involves 4 hidden layers) when compared to SVM and Logistic Regression. In future, if we have more data, ANN would certainly provide a much better accuracy score.